fix(service-worker): correctly determine client ID on navigation requests

gkalpak opened the pull request:
The ServiceWorker assigns an app-version to a each client to ensure that all subsequent requests for a client are served using the same app-version. The assignment is done based on the client ID.

Previously, the ServiceWorker would only try to read the client's ID off of the `FetchEvent`'s `clientId` property. However, for navigation requests the new client's ID will be set on [resultingClientId][1], while `clientId` will either be empty or hold the ID of the client where the request initiated from. See also related discussions in w3c/ServiceWorker#870 and w3c/ServiceWorker#1266.

In theory, this could lead to the navigation request (i.e. `index.html`) being served from a different app-version than the subsequent sub-resource requests (i.e. assets). In practice, the likelihood of this happening is probably very low though, since it would require the latest app-version to be updated between the initial navigation request and the first sub-resource request, which should happen very shortly after the navigation request.

This commit ensures that the correct client ID is determined even for navigation requests by also taking the `resultingClientId` property into account.

[1]: https://developer.mozilla.org/en-US/docs/Web/API/FetchEvent/resultingClientId

mary-poppins commented:
You can preview be5e838 at https://pr42607-be5e838.ngbuilds.io/.

mary-poppins commented:
You can preview b18287d at https://pr42607-b18287d.ngbuilds.io/.

petebacondarwin commented on packages/service-worker/worker/src/service-worker.d.ts:
```diff
@@ -60,8 +60,9 @@ type WindowClientState = 'hidden'|'visible'|'prerender'|'unloaded';
-  clientId: string|null;
   request: Request;
+  clientId?: string;
+  resultingClientId?: string;
```
Tiny nit:
Is it the case that one of `clientId` and `resultingClientId` must be defined?
If so, this could be enforced by using a union type here.

gkalpak replied:
To my understanding, these should actually both be defined (with one of them potentially being the empty string). But I am trying to align with the [webworker typings][1] provided by TS, because I am working on switching from our custom typings to the TS ones 🤷‍♂️ 

[1]: https://github.com/microsoft/TypeScript/blob/5afe42e14e61d7e4df5d75cc0022283711cb593a/lib/lib.webworker.d.ts#L201-L207

AndrewKushnir commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -609,7 +609,8 @@ export class Driver implements Debuggable, UpdateSource {
     // already be associated.
-    const clientId = event.clientId;
+    // (For navigation requests, we care about the `resultingClientId`.)
+    const clientId = event.resultingClientId || event.clientId;
```
nit: I'd propose adding a quick comment here saying that empty stings are not treated as valid values (thus the `||` works better than `??`) and we look at the `clientId` if the `resultingClientId` is an empty string, i.e. that it's intentional (and not a mistake of not handling empty strings).

gkalpak replied:
Added a comment.

gkalpak committed:
fix(service-worker): correctly determine client ID on navigation requests

The ServiceWorker assigns an app-version to a each client to ensure that
all subsequent requests for a client are served using the same
app-version. The assignment is done based on the client ID.

Previously, the ServiceWorker would only try to read the client's ID off
of the `FetchEvent`'s `clientId` property. However, for navigation
requests the new client's ID will be set on [resultingClientId][1],
while `clientId` will either be empty or hold the ID of the client where
the request initiated from. See also related discussions in
w3c/ServiceWorker#870 and w3c/ServiceWorker#1266.

In theory, this could lead to the navigation request (i.e. `index.html`)
being served from a different app-version than the subsequent
sub-resource requests (i.e. assets). In practice, the likelihood of this
happening is probably very low though, since it would require the latest
app-version to be updated between the initial navigation request and the
first sub-resource request, which should happen very shortly after the
navigation request.

This commit ensures that the correct client ID is determined even for
navigation requests by also taking the `resultingClientId` property into
account.

[1]: https://developer.mozilla.org/en-US/docs/Web/API/FetchEvent/resultingClientId

gkalpak committed:
fixup! fix(service-worker): correctly determine client ID on navigation requests

mary-poppins commented:
You can preview fac611a at https://pr42607-fac611a.ngbuilds.io/.
